src/quickgui/qgsquickhighlightsgnode.cpp

```diff
@@ -0,0 +1,168 @@
+{
+  Q_ASSERT( !geom.isMultipart() );
+
+  switch ( geom.type() )
+  {
+    case QgsWkbTypes::PointGeometry:
+    {
+      QVector<QgsPoint> points;
+      for ( auto it = geom.vertices_begin(); it != geom.vertices_end(); ++it )
+        points.append( *it );
+
+      if ( !points.isEmpty() )
+        appendChildNode( createPointGeometry( points.at( 0 ) ) );
+      break;
+    }
+
+    case QgsWkbTypes::LineGeometry:
+    {
+      QVector<QgsPoint> points;
+      for ( auto it = geom.vertices_begin(); it != geom.vertices_end(); ++it )
+        points.append( *it );
+
+      appendChildNode( createLineGeometry( points ) );
+      break;
+    }
+
+    case QgsWkbTypes::PolygonGeometry:
+    {
+      const QgsPolygon *poly = qgsgeometry_cast<const QgsPolygon *>( geom.constGet() );
+      if ( poly )
+        appendChildNode( createPolygonGeometry( *poly ) );
+      break;
+    }
+
+    case QgsWkbTypes::UnknownGeometry:
+    case QgsWkbTypes::NullGeometry:
+      break;
+  }
+}
+
+QSGGeometryNode *QgsQuickHighlightSGNode::createLineGeometry( const QVector<QgsPoint> &points )
+{
+  std::unique_ptr<QSGGeometryNode> node = qgis::make_unique< QSGGeometryNode>();
+  std::unique_ptr<QSGGeometry> sgGeom = qgis::make_unique< QSGGeometry>( QSGGeometry::defaultAttributes_Point2D(), points.count() );
+  QSGGeometry::Point2D *vertices = sgGeom->vertexDataAsPoint2D();
+
+  int i = 0;
+  for ( const QgsPoint &pt : points )
+  {
+    vertices[i++].set(
+      static_cast< float >( pt.x() ),
+      static_cast< float >( pt.y() )
+    );
+  }
+
+  sgGeom->setLineWidth( mWidth );
+  sgGeom->setDrawingMode( GL_LINE_STRIP );
+  node->setGeometry( sgGeom.release() );
+  node->setMaterial( &mMaterial );
+  node->setFlag( QSGNode::OwnsGeometry );
+  node->setFlag( QSGNode::OwnedByParent );
+  return node.release();
+}
+
+QSGGeometryNode *QgsQuickHighlightSGNode::createPointGeometry( const QgsPoint &point )
+{
+  std::unique_ptr<QSGGeometryNode> node = qgis::make_unique< QSGGeometryNode>();
+  std::unique_ptr<QSGGeometry> sgGeom = qgis::make_unique<QSGGeometry>( QSGGeometry::defaultAttributes_Point2D(), 1 );
+
+  QSGGeometry::Point2D *vertices = sgGeom->vertexDataAsPoint2D();
+  vertices[0].set(
+    static_cast< float >( point.x() ),
+    static_cast< float >( point.y() )
+  );
+  sgGeom->setDrawingMode( GL_POINTS );
+  sgGeom->setLineWidth( mWidth );
+
+  node->setGeometry( sgGeom.release() );
+  node->setMaterial( &mMaterial );
+  node->setFlag( QSGNode::OwnsGeometry );
+  node->setFlag( QSGNode::OwnedByParent );
+  return node.release();
+}
+
+QSGGeometryNode *QgsQuickHighlightSGNode::createPolygonGeometry( const QgsPolygon &polygon )
+{
+  QgsRectangle bounds = polygon.boundingBox();
+  QgsTessellator tes( bounds.xMinimum(), bounds.yMinimum(), false, false, false );
+  std::unique_ptr< QgsPolygon > p( qgsgeometry_cast< QgsPolygon * >( polygon.segmentize() ) );
+  tes.addPolygon( *p.get(), 0.0 );
+
+  std::unique_ptr<QgsMultiPolygon> triangles = tes.asMultiPolygon();
```

Comment:
Again - this whole block can be made much more efficient. Here you're calling asMultiPolygon(), which iterates over `tes.data()` and builds a multipolygon, and then you pick it back apart into separate vertices. 

You could instead get the data from `tes.data()` and then directly work with that (just like asMultiPolygon() does, but populating vertices directly)